### PR TITLE
fix MTP support on KDE Plasma and Dolphin

### LINKED ISSUE #151397: Android MTP issue in KDE: org.kde.kmtpd5: No such file or directory
### Describe the bug

As reported [here](https://discourse.nixos.org/t/android-file-transfert-mtp-support-has-always-been-half-broken-in-kde/16694), when I try to open my phone in Dolphin KDE, I get an error:
```
The file or the folder udi=/org/kde/solid/udev/sys/devices/pci0000:00/0000:00:14.0/usb1/1-4 does not exist.
```

If I run journalctl, I get a hint:
```
déc. 20 09:44:51 bestos dbus-daemon[6065]: [session uid=1000 pid=6065] Activated service 'org.kde.kmtpd5' failed: Failed to execute program org.kde.kmtpd5: No such file or directory
```

I'm thinking therefore that `kmtpd5` is maybe not packaged correctly. 

The full log being:
```
$ journalctl -f
déc. 20 09:44:50 bestos systemd[5577]: Started Dolphin - Gestionnaire de fichiers.
déc. 20 09:44:50 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/run/current-system/sw/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/applications.so'"
déc. 20 09:44:51 bestos xsession[13956]: kf.coreaddons: no metadata found in "/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so" "Failed to extract plugin meta data from '/nix/store/8b9xb90nzyi79fv2h70amkql6xyqidh2-plasma-workspace-5.22.5/lib/qt-5.15.2/plugins/kf5/kio/desktop.so'"
déc. 20 09:44:51 bestos xsession[6780]: qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 49375, resource id: 18937211, major code: 3 (GetWindowAttributes), minor code: 0
déc. 20 09:44:51 bestos xsession[6780]: qt.qpa.xcb: QXcbConnection: XCB error: 9 (BadDrawable), sequence: 49376, resource id: 18937211, major code: 14 (GetGeometry), minor code: 0
déc. 20 09:44:51 bestos dbus-daemon[6065]: [session uid=1000 pid=6065] Activating service name='org.kde.kmtpd5' requested by ':1.59' (uid=1000 pid=14074 comm="mtp.so [kdeinit5] mtp local:/run/user/1000/klaunch" label="kernel")
déc. 20 09:44:51 bestos dbus-daemon[6065]: [session uid=1000 pid=6065] Activated service 'org.kde.kmtpd5' failed: Failed to execute program org.kde.kmtpd5: No such file or directory
```

(I'm also not sure what is this `Failed to extract plugin meta data`, but it does not look super nice)

### Steps To Reproduce
1. Install KDE/Plasma
2. Plug an android phone
3. Open it in dolphin

### Notify maintainers
@NixOS/qt-kde 

### Potentially linked

It may be potentially linked with https://github.com/NixOS/nixpkgs/issues/145824 but there they mention about gvfs (I don't think I need gvfs for Dolphin to use mtp). So I created instead a separate issue.

### NixOs version

Running `a4bf4434570`.

### PR DESCRIPTION
**Pull request by olegkapitonov**
###### Motivation for this change

MTP support on KDE Plasma and Dolphin is broken, this change will fix it.

Fixes #151397

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hqurve** [comment]
Works for me. 

To test: i added the following overlay

```pself: psuper: 
let
  libsForQt5Overlay = libsForQt5: libsForQt5.overrideScope' (lself: lsuper: 
    let
      kdeGear = lsuper.kdeGear.overrideScope' (self: super: {
        kio-extras = super.kio-extras.overrideAttrs (old: {
          postInstall = ''
            ${old.postInstall or ""}
            sed 's,Exec=.*,Exec=${lsuper.kio}/libexec/kf5/kiod5,' "$out/share/dbus-1/services/org.kde.kmtpd5.service"
          '';
        });
      });
    in (kdeGear // { 
      inherit kdeGear;
      kdeApplications = kdeGear;
    })
  );
in {
  # wasnt too sure which to apply it to, so I applied it to both
  libsForQt5 = libsForQt5Overlay psuper.libsForQt5;
  plasma5Packages = libsForQt5Overlay psuper.plasma5Packages;
}
# is there a simpler way to apply an overlay to qt5packages?
```
and manually added `services.udev.packages = [ pkgs.libmtp.out ];` to my system configuration.
Then ran nixos-rebuild switch and upon plugging in my phone, it just worked™

**Comment by olegkapitonov** [comment]
Hello, @hqurve!

Your log shows that this part of my fix

```
postInstall = ''
  sed 's,Exec=.*,Exec=${lsuper.kio}/libexec/kf5/kiod5,' "$out/share/dbus-1/services/org.kde.kmtpd5.service"
'';
```

somehow didn't apply in your configuration. Maybe something wrong with your overlay.
Please check `share/dbus-1/services/org.kde.kmtpd5.service` file in `kio-extras` derivation and ensure that `Exec=` points into the `kio` derivation, not `kio-extras`.

**Comment by hqurve** [comment]
Indeed, `/run/current-system/share/dbus-1/services/org.kde.kmtpd5.service` contains kio-extras instead of kio. Also, the file in the derivation itself seems  to not have changed (despite requiring a rebuild)

File: pkgs/applications/kde/kio-extras.nix [hunk 1]
```diff
@@ -19,5 +19,11 @@ mkDerivation {
     kpty syntax-highlighting libmtp libssh openexr openslp
     phonon qtsvg samba solid gperf
   ];
+
+  # org.kde.kmtpd5 DBUS service launches kiod5 binary from kio derivation, not from kio-extras
+  postInstall = ''
+    sed 's,Exec=.*,Exec=${kio}/libexec/kf5/kiod5,' "$out/share/dbus-1/services/org.kde.kmtpd5.service"
```
Comment (hqurve) [inline pkgs/applications/kde/kio-extras.nix:25]:
```suggestion
    sed -i 's,Exec=.*,Exec=${kio}/libexec/kf5/kiod5,' "$out/share/dbus-1/services/org.kde.kmtpd5.service"
```

**Review by hqurve — COMMENTED** [review]
Figured it out, you forgot to sed inplace

**Comment by hqurve** [comment]
Reapplied with the new change, nothing worked. However, after I rebooted the machine, it works!

**Comment by hqurve** [comment]
Also, I am not sure if this is related to this however, simply clicking "mount and open" from the widget in the system tray navigates me to my home-directory (I assume its actually whatever behaviour occurs when you open dolphin with no arguments. Which in my case is my home-directory). In order to access the device I have to click the device in dolphin's sidepanel

**Comment by olegkapitonov** [comment]
Thanks a lot for fix, @hqurve!

When I plug my phone, unlock the screen, select "File transfer" option - the widget in the system tray navigates to the internal memory of the phone. All works for me.

I think this may be another KDE issue, not related to this subject.

File: pkgs/applications/kde/kio-extras.nix (hunk 1 above)
Comment (ElXreno) [inline pkgs/applications/kde/kio-extras.nix:25]:
```suggestion
    substituteInPlace $out/share/dbus-1/services/org.kde.kmtpd5.service \
      --replace Exec=$out Exec=${kio}
```

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build haruna dolphin

**Comment by peterhoeg** [comment]
@GrahamcOfBorg test plasma5

**Comment by FliegendeWurst** [comment]
Can confirm this also fixes the issue on NixOS 21.11.

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build haruna dolphin
@GrahamcOfBorg test plasma5

**Comment by cirno-999** [comment]
Issue still persists for me on 21.11.

[cirno@berlin:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.93, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5.0pre20211206_d1aaa7e`
 - channels(root): `"nixos-21.11.335526.6c4b9f1a2fd, unstable-22.05pre334855.29d1f6e1f62"`
 - channels(cirno): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

image

**Comment by hqurve** [comment]
@cirno-999 I believe the issue you are experiencing is not related to mtp but with kdeconnect (as suggested by your screenshot). More specifically, I believe it is [this](https://bugs.kde.org/show_bug.cgi?id=443155) issue.

Currently, i am on nixos-unstable (945ec499041db73043f745fad3b2a3a01e826081 ~1 week old) and kde connect works. However, previously (I dont remember when), i was getting a similar issue.

From the previously linked bug report, I believe the issue is to do with a desync between the kdeconnect on your phone and ssh used by the computer

**Comment by cirno-999** [comment]
@hqurve yes, you are right. I mistook the protocols. Now I removed the kdeconnect completely, but when trying to connect with MTP, the device doesn't appear anywhere (not under mtp:/, no errors shown either). I'm happy to provide any debug info but I've not seen any errors. The device works under W*ndows too.

**Comment by hqurve** [comment]
Firstly, would you mind creating a separate issue for this and linking it in a comment below? (a PR isnt the right place to discuss this).

In it, I think you should see what messages are produced in the journal. Specifically, plug the device in and then copy the relevant journal messages by running `journalctl --since="5m" --no-pager | <your favorite editor>`. You should be looking for a message about a new usb device similar to those I posted above.

And of course, put in all the information which the issue template asks (if possible).

**Pull request merged**